[Web share] canShare() validation returns incorrect result when sharing valid files but invalid URL

marcoscaceres opened the pull request:
#### 44186059d58ae7dfeb5baefbca62a757d7e58fdd
<pre>
[Web share] canShare() validation returns incorrect result when sharing valid files but invalid URL
<a href="https://bugs.webkit.org/show_bug.cgi?id=242502">https://bugs.webkit.org/show_bug.cgi?id=242502</a>

Reviewed by Youenn Fablet.

* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/web-share/canShare.https-expected.txt:
* LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/web-share/canShare.https-expected.txt:
* Source/WebCore/page/Navigator.cpp:
(WebCore::shareableURLForShareData):
(WebCore::Navigator::canShare):

Canonical link: <a href="https://commits.webkit.org/252403@main">https://commits.webkit.org/252403@main</a>
</pre>

youennf commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -137,15 +137,25 @@ bool Navigator::canShare(Document& document, const ShareData& data)
+
+#if ENABLE(FILE_SHARE)
+    bool hasTitleTextOrURL = !data.title.isNull() || !data.text.isNull() || !data.url.isNull();
+    if (!hasTitleTextOrURL && !hasShareableFiles)
```
Isn't that redundant with line 146 check?
And why is this check controlled by #if ENABLE(FILE_SHARE)?

marcoscaceres replied:
yeah, true... I was coding it to match the spec text. I'll fold this up.

youennf commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -137,15 +137,25 @@ bool Navigator::canShare(Document& document, const ShareData& data)
+    if (!data.url.isNull() && !shareableURLForShareData(document, data))
+        return false;
+
+    return true;
```
Could be a one liner:
return data.url.isNull() || shareableURLForShareData(document, data);

marcoscaceres replied:
As above. Will fix.

marcoscaceres committed:
[Web share] canShare() validation returns incorrect result when sharing valid files but invalid URL
https://bugs.webkit.org/show_bug.cgi?id=242502

Reviewed by Youenn Fablet.

* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/web-share/canShare.https-expected.txt:
* LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/web-share/canShare.https-expected.txt:
* Source/WebCore/page/Navigator.cpp:
(WebCore::shareableURLForShareData):
(WebCore::Navigator::canShare):

Canonical link: https://commits.webkit.org/252403@main
